{
  "cells": [
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "view-in-github",
        "colab_type": "text"
      },
      "source": [
        "<a href=\"https://colab.research.google.com/github/fmaguire/pha4gecon_hAMRonization/blob/master/pha4gecon_hAMRonization_demo.ipynb\" target=\"_parent\"><img src=\"https://colab.research.google.com/assets/colab-badge.svg\" alt=\"Open In Colab\"/></a>"
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "__I4n8V7-R0B"
      },
      "source": [
        "\n",
        "##**Google Colab Based Demo for hAMRonization**\n",
        "\n",
        "This is based on a notebook by Su Datt Lam and Hui-min Neoh at Universiti Kebangsaan Malaysia.\n",
        "\n",
        "hAMRonization is a tool which converts AMR detection tool outputs (such as those generated by resfinder4, amrfinderplus, RGI etc) to a standardized dat specification and report (i.e., hAMRonization specification format).\n",
        "\n",
        "This demo will show you how to run hAMRonization in your browser using google colab. **Note**: Don't include personal health information when using google servers unless you have checked with your privacy office!\n",
        "\n",
        "<img src=\"https://github.com/pha4ge/hAMRonization/raw/master/docs/overview_figure.png?raw=true\" height=\"600\" align=\"centre\" style=\"height:400px\">\n",
        "\n",
        "For more details regarding the software, checkout the [hAMRonization GitHub](https://github.com/pha4ge/hAMRonization).\n",
        "\n",
        "# Instructions\n",
        "Please press the \"Connect\" button found on the upper right of this page to connect to a Google server. The pipeline consists of 4 steps. The currently running steps is indicated by a circle with a stop sign next to it:\n",
        "  1.   Installing hAMRonization\n",
        "  2.   Running your sample through both RGI online\n",
        "  3.   hAMRonizing the results from these with hAMRonization\n",
        "  4.   Generating a combined report using hAMRonization"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "metadata": {
        "id": "FK2RXzcaVqy5"
      },
      "outputs": [],
      "source": [
        "#@title 1a. Install the hAMRonization software { vertical-output: true }\n",
        "\n",
        "#@markdown Please execute this cell by pressing the *Play* button on the left.\n",
        "!pip install hAMRonization"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "metadata": {
        "id": "sADc0u-iV0p2",
        "cellView": "form"
      },
      "outputs": [],
      "source": [
        "#@title 1b. Test the hAMRonization installation { vertical-output: true }\n",
        "\n",
        "!hamronize rgi -h"
      ]
    },
    {
      "cell_type": "markdown",
      "source": [
        "## 2. Predict AMR genes in a genome using RGI\n",
        "\n",
        "You can use any bacterial genome you have in fasta format but here is [an example](https://raw.githubusercontent.com/fmaguire/pha4gecon_hAMRonization/master/enterobacter_assembly.fasta) you can download for the demo by just saving/hitting ctrl-s after opening that link.\n",
        "\n",
        "Now go to CARD's [RGI](https://card.mcmaster.ca/analyze) and predict AMR genes in your genome by uploading your fasta file then clicking \"submit\". Wait until it completes and download your results using the botton at the top.  You will then have to extract the .txt file from the .tar.gz that downloads."
      ],
      "metadata": {
        "id": "q9vIUsJZl6QC"
      }
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "metadata": {
        "id": "4AoMrTz2WKIT"
      },
      "outputs": [],
      "source": [
        "#@title 3a. hAMRonization: Upload your RGI results in txt format (e.g., DATE-TIME.XXX_enterobacter_assembly.fasta.txt ) { vertical-output: true }\n",
        "\n",
        "#@markdown Please execute this cell by pressing the *Play* button on the left. Then, the upload file button will then appeared.\n",
        "\n",
        "#@markdown Upload the file by pressing \"Choose Files\". You are allowed to only 1 file.\n",
        "\n",
        "from google.colab import files\n",
        "\n",
        "uploaded = files.upload()\n",
        "\n",
        "for fn in uploaded.keys():\n",
        "  print('User uploaded file \"{name}\"'.format(name=fn))"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "metadata": {
        "id": "bt7kcmt2xko1"
      },
      "outputs": [],
      "source": [
        "#@title 3b. Provide the parameters and run hAMRonization on this RGI output { vertical-output: true }\n",
        "#@markdown\n",
        "#@markdown ---\n",
        "filename = '2023-10-26_10:55:47.222_enterobacter_assembly.fasta.txt' #@param {type: \"string\"}\n",
        "#@markdown  - the name of the file you uploaded. Please include the extension name (for example, .txt).\n",
        "program = 'rgi'  #@param {type: \"string\"}\n",
        "#@markdown  - program supported: abricate; amrfinderplus; ariba; rgi; resfinder; resfinder4; srax; deeparg; kmerresistance; srst2; staramr; csstar; amrplusplus; resfams; groot\n",
        "input_filename = 'enterobacter_assembly' #@param {type: \"string\"}\n",
        "#@markdown  - Sequence entry\n",
        "analysis_software_version = '6.0.2' #@param {type: \"string\"}\n",
        "#@markdown  - AMR software version (get this from the card.mcmaster.ca/analyze rgi-main page).\n",
        "reference_database_version = '3.2.8' #@param {type: \"string\"}\n",
        "#@markdown  - Database version (get this from the card.mcmaster.ca/analyze rgi-main page).\n",
        "format = 'tsv' #@param {type: \"string\"}\n",
        "#@markdown  - tsv; json; interactive\n",
        "output_filename = 'demo_rgi.txt' #@param {type: \"string\"}\n",
        "#@markdown  - Filename of output file.\n",
        "#@markdown ---\n",
        "\n",
        "!hamronize {program} --input_file_name {input_filename} --analysis_software_version {analysis_software_version} --reference_database_version {reference_database_version} {filename} --format {format} --output {output_filename}\n",
        "\n",
        "import os\n",
        "\n",
        "path_name = output_filename\n",
        "rgi_output = output_filename\n",
        "if os.path.exists(path_name):\n",
        "  print(\"Harmonization complete for \"+ output_filename +\".\")\n"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "metadata": {
        "id": "XmVdHZ5Q9KHs"
      },
      "outputs": [],
      "source": [
        "#@title 3c. Check and download the hAMRonization output { vertical-output: true }\n",
        "\n",
        "#@markdown If you are having issues downloading the result archive, try disabling your adblocker and run this cell again. If that fails click on the little folder icon to the left, navigate to file: output_filename, right-click and select \"Download\".\n",
        "\n",
        "#@markdown The content of the output file is as below:\n",
        "\n",
        "#@markdown ---\n",
        "\n",
        "with open(path_name, 'r') as f:\n",
        "    print(f.read())\n",
        "\n",
        "files.download(output_filename)"
      ]
    },
    {
      "cell_type": "code",
      "source": [
        "#@title 4. Summarize and download your hAMRonization results { vertical-output: true }\n",
        "!hamronize summarize -t interactive {rgi_output} -o hamronized_summary.html\n",
        "!hamronize summarize -t tsv {rgi_output} -o hamronized_summary.tsv\n",
        "\n",
        "files.download(\"hamronized_summary.html\")\n",
        "files.download(\"hamronized_summary.tsv\")"
      ],
      "metadata": {
        "id": "j3YpMSH6sKUs"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "markdown",
      "source": [
        "You can then open \"hamronized_summary.html\" in your browser to explore your results in this standardised format!\n",
        "\n",
        "If you have more time - try installing and running your genome through [AMRFinderPlus](https://github.com/ncbi/amr) downloading those results and hamronize them.\n",
        "\n",
        "You can then summarize the RGI and AMRFinderPlus results together to compare them."
      ],
      "metadata": {
        "id": "F6h0IBF_sxz5"
      }
    }
  ],
  "metadata": {
    "colab": {
      "provenance": [],
      "include_colab_link": true
    },
    "kernelspec": {
      "display_name": "Python 3",
      "name": "python3"
    },
    "language_info": {
      "name": "python"
    }
  },
  "nbformat": 4,
  "nbformat_minor": 0
}